### ISSUE TITLE
perf: Float32 viability experiment for JaxEngine harmonic force fields

### ISSUE THREAD

**Issue by ericchansen** [ericchansen/q2mm#178]
## Summary

Test whether float32 is viable for JaxEngine's harmonic-only force fields. If it is, consumer GPUs like the RTX 5090 unlock 64× more throughput (104.8 vs 1.6 TFLOPS).

## Background

q2mm currently forces `jax_enable_x64=True` globally in `_jax_common.py`. [GPU benchmarks](https://ericchansen.github.io/q2mm/benchmarks/gpu/) show CPU beating GPU by 1.6–5.5×, and the float64 penalty on consumer GPUs is the dominant factor.

Numerical analysis suggests float64 may not be necessary for the current harmonic-only JaxEngine:

- The unit conversion (kcal/mol/Å² → Hartree/Bohr²) is **linear** — it does not change relative precision
- For harmonic terms (`E = k(r − r₀)²`), the Hessian is exactly `2k` — no catastrophic cancellation in the autodiff tape
- To resolve 1 cm⁻¹ at a 1000 cm⁻¹ mode requires ~8% relative precision; float32 provides 1.2 × 10⁻⁷ — **5 orders of magnitude** more than needed

Float64 would be needed for:
- VdW terms (1/r¹² − 1/r⁶ near-cancellation)
- Morse potentials, cross-terms
- Very soft modes (~10 cm⁻¹)

## Experiment

1. Run CH₃F and rh-enamide benchmarks with `jax_enable_x64=False`
2. Compare computed frequencies against float64 baseline
3. Measure GPU throughput improvement
4. Document results in `docs/benchmarks/gpu.md`

## Success criteria

- [ ] Frequency comparison table (float32 vs float64) for both systems
- [ ] GPU benchmark numbers with float32
- [ ] Clear recommendation: viable / not viable / conditionally viable
- [ ] If viable, document which force field terms are safe in float32

## Related

- #176 (end-to-end differentiable loss — this experiment informs Phase 0)
- #155 (GPU acceleration umbrella)

**Comment by ericchansen** [comment]
## Mixed-Precision Experiment Results

Follow-up to the full-float32 experiment. Tests whether computing the Hessian in float32 (GPU-friendly) and then casting to float64 for eigendecomposition recovers precision.

### Methodology

Three paths compared on the same molecules:
1. **Full f64**: `jax.hessian` in float64 → `eigvalsh` in float64 (baseline)
2. **Full f32**: `jax.hessian` in float32 → `eigvalsh` (numpy, inherently f64 but input is f32-precision)
3. **Mixed**: `jax.hessian` in float32 → **explicit cast to f64** → `eigvalsh` in float64

All runs in ci-jax container (CPU), isolating precision effects from hardware.

### Results

**CH₃F (5 atoms, 15 DOF):**
```
Mixed precision:  Max Δ = 0.000160 cm⁻¹  ✅ PASS
Hessian element diff: max=1.1e-04, rel_max=8.0e-08
```

**Rh-enamide mol0 (36 atoms, 108 DOF):**
```
Mixed precision:  Max Δ = 0.443 cm⁻¹   ❌ FAIL (0.1 cm⁻¹ threshold)
Hessian element diff: max=18.7, rel_max=1.9e-06

Worst deviations (all in soft modes):
  freq[35]:  f64=65.07   mixed=65.51   Δ=0.443 cm⁻¹
  freq[38]:  f64=85.59   mixed=85.77   Δ=0.176 cm⁻¹
  freq[37]:  f64=79.52   mixed=79.69   Δ=0.167 cm⁻¹
```

### Analysis

Mixed precision is **better** than full-f32 (0.44 vs 0.78 cm⁻¹) but still fails the 0.1 cm⁻¹ threshold. The bottleneck is the Hessian computation itself — `jax.hessian` in float32 has ~10⁻⁶ relative element errors, which propagate into eigenvalues of soft modes (60–86 cm⁻¹).

The eigendecomposition is NOT the problem — casting to f64 before `eigvalsh` is accurate. It's the input Hessian that's too noisy.

### Implications

The 0.1 cm⁻¹ threshold is strict. With a relaxed threshold:
- **0.5 cm⁻¹**: mixed precision passes (barely) → 64× GPU throughput on consumer cards
- **1.0 cm⁻¹**: comfortable margin
- **Small molecules (≤10 atoms)**: always passes at 0.1 cm⁻¹

A future adaptive strategy could use float32 for early optimisation cycles and switch to float64 for refinement.

Full results and methodology documented in [`docs/benchmarks/gpu.md`](https://ericchansen.github.io/q2mm/benchmarks/gpu/#float32-viability-test).
Scripts: `scripts/float32_experiment.py`, `scripts/mixed_precision_experiment.py`.
